packages/flutter/lib/src/cupertino/sliding_segmented_control.dart

```diff
@@ -485,16 +503,26 @@ class _SegmentedControlState<T extends Object> extends State<CupertinoSlidingSeg
   // them from interfering with the active drag gesture.
   bool get isThumbDragging => _startedOnSelectedSegment ?? false;
 
-  // Converts local coordinate to segments. This method assumes each segment has
-  // the same width.
+  // Converts local coordinate to segments. If `widget.proportionalWidth` is
+  // false, this method assumes each segment has the same width; if it is true,
+  // this method assumes segments have their own width based on their contents.
   T segmentForXPosition(double dx) {
-    final RenderBox renderBox = context.findRenderObject()! as RenderBox;
+    final TextDirection textDirection = Directionality.of(context);
+    final BuildContext currentContext = segmentedControlRenderWidgetKey.currentContext!;
+    final _RenderSegmentedControl<T> renderBox = currentContext.findRenderObject()! as _RenderSegmentedControl<T>;
+
+    if (widget.proportionalWidth) {
```

Comment:
I think you should be able to ask the render object regardless of whether `proportionalWidth` is set?

Reply:
It looks like the original implementation always returns an index in [0, leng - 1). `getSegmentIndex` should probably be renamed to `getClosestSegmentIndex` and always return a valid index?

Reply:
uber nit: the RTL reordering happens at the widget level so consider doing the RTL thing in the state.

Reply:
Updated based on the comments:)